### PR TITLE
Make HAProxy work as an http proxy instead of a tcp proxy.

### PR DESCRIPTION
**Pull request by ereslibre**
This allows us to add fine-grained timeouts depending on the endpoint being
accessed or with what parameters (e.g. /log?follow=true should have no
timeout as happens on the apiserver). /exec is another example, but in this
case the protocol is upgraded to spdy.

Depends on https://github.com/kubic-project/caasp-container-manifests/pull/164
Depends on https://github.com/kubic-project/automation/pull/377

Fixes: bsc#1071994

### CODE REVIEW

**Comment by ereslibre** [comment]
I will resume this work when I have more time. This is interesting for the conformance tests as we wouldn't have to do tricks in order to find out whether the tests have finished yet or not. Also, it's fixing a bug with `kubectl exec` (not this specific PR at the moment, but I plan to improve it to also cover the `kubectl exec` case) in which the connection timeouts on the proxy.

**Comment by ereslibre** [comment]
[Bug 1071994 - kubectl run --attach exits early](https://trello.com/c/nNNdxJAn/704-bug-1071994-kubectl-run-attach-exits-early)

**Comment by ereslibre** [comment]
The reason why this is failing on CI is because when rebooting the admin node to update, `admin-node-setup.sh` runs on the admin node on bootup writing old manifests, so haproxy cannot start (interrupting the real update of the cluster as Velum will be down):

https://github.com/kubic-project/caasp-container-manifests/blob/2ce28547afef995aabe6fc132f585d79ece3009a/admin-node-setup.sh#L33-L48

Basically the issue is that we have new `haproxy` configuration with old manifests (with old mounts), what means that `haproxy` cannot find on the new locations the configuration as we copied over the old manifests with the old mounts.

File: salt/top.sls [hunk 1]
```diff
@@ -9,6 +9,7 @@ base:
     - proxy
     - rebootmgr
     - transactional-update
+    - kube-apiserver-proxy
```
Comment (inercia) [inline salt/top.sls]: Do we need this new top-level dir? It looks to me the code in  `salt/kube-apiserver-proxy/init.sls` (it is just a certificca generation) could be just moved to `salt/haproxy/init.sls`

File: salt/haproxy/haproxy.cfg.jinja [hunk 2]
```diff
@@ -9,58 +9,77 @@ global
...
+        http-request set-header X-Remote-User  %{+Q}[ssl_c_s_dn(cn)] if { ssl_c_used ssl_c_verify 0 }
+        http-request set-header X-Remote-Group %{+Q}[ssl_c_s_dn(o)] if { ssl_c_used ssl_c_verify 0 }
+        # If no certificate is passed, or if it's invalid, remove the auth headers
+        http-request del-header X-Remote-User  unless { ssl_c_used ssl_c_verify 0 }
+        http-request del-header X-Remote-Group unless { ssl_c_used ssl_c_verify 0 }
+        http-request del-header X-Remote-Extra unless { ssl_c_used ssl_c_verify 0 }
+        acl streaming_logs url_reg [?&]follow=true
+        use_backend no-timeout-backend if streaming_logs
+        acl interactive_session url_reg [?&]tty=true
+        use_backend no-timeout-backend if interactive_session
```
Comment (inercia) [inline salt/haproxy/haproxy.cfg.jinja]: (I'm not an haproxy expert, so) is this setting the `haproxy` as the SSL termination? Will the `haproxy`<->`master` traffic be in plain HTTP?

**Comment by kiall** [comment]
> The reason why this is failing on CI is because when rebooting the admin node to update, admin-node-setup.sh runs on the admin node on bootup writing old manifests, so haproxy cannot start (interrupting the real update of the cluster as Velum will be down):

So - do we need a matching change to the haproxy manifest in caasp-container-manifests?

File: salt/top.sls (hunk 1 above)
Comment (ereslibre) [reply salt/top.sls]: Yes, I'll move it.

File: salt/haproxy/haproxy.cfg.jinja (hunk 2 above)
Comment (ereslibre) [reply salt/haproxy/haproxy.cfg.jinja]: It's TLS terminated.

**Comment by ereslibre** [comment]
> So - do we need a matching change to the haproxy manifest in caasp-container-manifests?

No. My idea was to fall back to `/etc/pki/ca.crt` on the `haproxy` config, as we cannot use `/etc/pki/trust/anchors/SUSE_CaaSP_CA.crt` (`ssl:ca_file`) on caasp-container-manifests because on the first run of the admin node this file doesn't exist and I don't want to add more glue to `caasp-container-manifests`.

The fix should be trivial, but I just set this aside despite almost done because we have a number of other pressing issues :(

**Comment by ereslibre** [comment]
That being said, updating `caasp-container-manifests` might be needed, but not for fixing this issue in particular, for the certificate I'd just fall back to `/etc/pki/ca.crt`.

File: salt/haproxy/haproxy.cfg.jinja (hunk 2 above)
Comment (inercia) [reply salt/haproxy/haproxy.cfg.jinja]: Ok, TLS, but I mean that the `haproxy`<->`master` communication will be in plain HTTP, right?

Comment (ereslibre) [reply salt/haproxy/haproxy.cfg.jinja]: No, that communication happens also over TLS.

**Comment by ereslibre** [comment]
This won't pass on CI until https://github.com/kubic-project/caasp-container-manifests/pull/164 is on the image.

**Comment by ereslibre** [comment]
Rebased and conflicts fixed. Will address comments at a later time.

**Comment by ereslibre** [comment]
This also depends on https://github.com/kubic-project/automation/pull/377

**Comment by ereslibre** [comment]
Please, do not merge: figuring out why the two PR's together aren't passing in Jenkins: https://github.com/kubic-project/jenkins-library/pull/170

**Comment by ereslibre** [comment]
Passing with: https://github.com/kubic-project/jenkins-library/pull/170

**Comment by ereslibre** [comment]
@flavio @kiall: We need to force merge this one. Passing on https://github.com/kubic-project/jenkins-library/pull/170

**Pull request merged**